src/main/webapp/availability.jsp

```diff
@@ -0,0 +1,56 @@
+<html>
+  <head>
+    <meta charset="UTF-8">
+    <title>Availability</title>
+    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min.css"
+      integrity="sha384-Vkoo8x4CGsO3+Hhxv8T/Q5PaXtkKtu6ug5TOeNV6gBiFeWPGFN9MuhOf23Q9Ifjh" crossorigin="anonymous">
+    <link rel="stylesheet" href="global.css">
+    <link rel="stylesheet" href="availability.css">
+    <script src="https://code.jquery.com/jquery-3.4.1.slim.min.js"
+      integrity="sha384-J6qa4849blE2+poT4WnyKhv5vZF5SrPo0iEjwBvKU7imGFAV0wwj1yYfoRSJoZ+n" crossorigin="anonymous"></script>
+    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js"
+      integrity="sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo" crossorigin="anonymous"></script>
+    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js"
+      integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6" crossorigin="anonymous"></script>
+    <script src="availability.js"></script>
+    <script src="logout.js"></script>
+  </head>
+  <body onload="onAvailabilityLoad()">
+    <div id="content">
+      <nav class="navbar navbar-expand-lg navbar-light bg-light">
+        <span class="navbar-brand mb-0 h1">InterviewMe</span>
+        <button class="navbar-toggler" type="button" data-toggle="collapse" 
+          data-target="#navbarNavAltMarkup" aria-controls="navbarNavAltMarkup" 
+          aria-expanded="false" aria-label="Toggle navigation">
+          <span class="navbar-toggler-icon"></span>
+        </button>
+        <div class="collapse navbar-collapse" id="navbarNavAltMarkup">
+          <div class="navbar-nav">
+            <a class="nav-item nav-link" href="index.html">Home</a>
+            <a class="nav-item nav-link" href="profile.html">Profile</a>
+            <a class="nav-item nav-link" href="scheduled-interviews.html">Events</a>
+            <a class="nav-item nav-link active" href="#">Availability</a>
+            <a class="nav-item nav-link" href="search-interview.html">Scheduling</a>
+            <!--TODO: Move this to the right of the navbar-->
+            <a class="nav-item nav-link" href="#" id="login-tab">Logout</a>
+          </div>
+        </div>
+      </nav>
+      <br>
+      <h1 class="text-center">Availability</h1>
+      <h5 class="text-center lead">Update your availability to be an Interviewer for this week below:</h5>
+      <br>
+      <!--TODO: Include Timezone consideration; Check inputs for correctness: end is after start, at least an hour long;
+         Add pagination; Figure out how to update the dates (dynamically); store and show inputted data; Allow for more than one
+         time range per day (have a plus button)-->
+         
+      <!--TODO: Put entire table in a JSP file that will then be included here (meaning that this will need to be a jsp file too).-->
+      <jsp:include page = "availabilityTable.jsp" flush = "true"/>
```

Comment:
I'm thinking based on our previous discussions that you won't want to include this here. Instead, you'll do something onLoad of the page, that inserts pre-rendered jsp (from the server) into a placeholder div.

Reply:
(That's not how jsp is usually used, to be clear, just how we're discussing using it here)

Reply:
So to be more explicit, my suggestion is:

1. Replace this with an empty placeholder div (with some distinct id or class).
2. Add or update your onLoad function to include "loadAvailabilityTable"
3. In loadAvailabilityTable, make a request (i.e. fetch()) to a backend endpoint which returns the calendar HTML. Temporary placeholder is fine (just return some pregenerated static string)

---

In a followup, implement the backend endpoint to generate the HTML from jsp.

Reply:
To attempt to follow the flow all the way through: 
1. onLoad() calls a JavaScript Function that gets the current time zone.
2. The time zone is fed into a get request to the AvailabilityTimeSlotsServlet.
3. Here the time zone is used to create a List of AvailabilityTimeSlot objects.
4. This list is accessed by the availabilityTable JSP file to create the table in HTML.
5. JSP is fed back to the AvailabilityTimeSlotsServlet (not currently clear on how).
6. Servlet sends HTML response to JavaScript.
7. JavaScript does something like tableDiv.innerHTML = response.

Reply:
> 1. onLoad() calls a JavaScript Function that gets the current time zone.
> 2. The time zone is fed into a get request to the AvailabilityTimeSlotsServlet.
> 3. Here the time zone is used to create a List of AvailabilityTimeSlot objects.
 
This sounds right to me.

> 4. This list is accessed by the availabilityTable JSP file to create the table in HTML.
> 5. JSP is fed back to the AvailabilityTimeSlotsServlet (not currently clear on how).

I would word this differently. I believe you should be able to access a JSP "renderer", so what I would expect is that you configure the "renderer" with the jsp file (or files), then ask it to render one template with the required parameters. This results in a string containing the HTML.

Let me look into the APIs that are available, I'll coordinate with you on this in chat.

> 6. Servlet sends HTML response to JavaScript.
> 7. JavaScript does something like tableDiv.innerHTML = response.

Yep.

Reply:
test

Reply:
I looked into this further and didn't find an easy to use renderer, and realized that theres actually a more conventional workflow here.

(All of the below was sent to Meagan before, duplicating it here for the historical record)

---

You mentioned:
> There is a current plan that involves fetching a jsp file from JavaScript and using java objects within the jsp file to generate the html.

I think this approach is right. Basically the page loads, runs onLoad (on body), which does a fetch. That javascript function runs something like this:

```javascript
onLoad = () => {
  // TODO: to be implemented
  tz = getBrowserTimezone();
  // TODO: to be implemented
  outputDiv = getAppointmentDiv();
  fetch("/appointmentsWidget?tz="+encodeURIComponent(tz))
    .then(getAppointmentWidgetHandler(outputDiv))
}

getAppointmentWidgetHandler = (outputDiv) => {   
  // Return a callback suitable for onreadystatechange.
  return (response) => {
    // TODO: handle errors
    // ...

    // Get the contents of the response, which should be HTML.
    contents = ""; // TODO: response.???
    outputDiv.innerHTML = contents;
  }
}
```

With that, you just need to implement an appointmentWidget servlet that works like a regular jsp page (follow tutorials for that). That jsp page should not include the navbar or any of that other content... Just the appointment slot widget.

You could break this into two independent problems. One, to make the jsp widget work (just put it as a navbar link temporarily and use a hardcoded testing timezone in the URL). Next, do what I did above to make the request as part of the appointments page load, and then replace the text.